aten/src/ATen/native/cpu/UpSampleKernelAVXAntialias.h

```diff
@@ -342,376 +365,602 @@ void upsample_avx_bilinear_uint8(
-
-      source = _mm256_inserti128_si256(
-          _mm256_castsi128_si256(_mm_loadu_si128((__m128i*)&lineIn0[x + xmin])),
-          _mm_loadu_si128((__m128i*)&lineIn1[x + xmin]),
-          1);
-      // clang-format off
-      pix = _mm256_shuffle_epi8(source, _mm256_set_epi8(
-        -1,7, -1,3, -1,6, -1,2, -1,5, -1,1, -1,4, -1,0,
-        -1,7, -1,3, -1,6, -1,2, -1,5, -1,1, -1,4, -1,0));
-      sss0 = _mm256_add_epi32(sss0, _mm256_madd_epi16(pix, mmk0));
-      pix = _mm256_shuffle_epi8(source, _mm256_set_epi8(
-        -1,15, -1,11, -1,14, -1,10, -1,13, -1,9, -1,12, -1,8,
-        -1,15, -1,11, -1,14, -1,10, -1,13, -1,9, -1,12, -1,8));
-      sss0 = _mm256_add_epi32(sss0, _mm256_madd_epi16(pix, mmk1));
-
-      source = _mm256_inserti128_si256(
-          _mm256_castsi128_si256(_mm_loadu_si128((__m128i*)&lineIn2[x + xmin])),
-          _mm_loadu_si128((__m128i*)&lineIn3[x + xmin]),
-          1);
-      pix = _mm256_shuffle_epi8(source, _mm256_set_epi8(
-        -1,7, -1,3, -1,6, -1,2, -1,5, -1,1, -1,4, -1,0,
-        -1,7, -1,3, -1,6, -1,2, -1,5, -1,1, -1,4, -1,0));
-      sss1 = _mm256_add_epi32(sss1, _mm256_madd_epi16(pix, mmk0));
-      pix = _mm256_shuffle_epi8(source, _mm256_set_epi8(
-        -1,15, -1,11, -1,14, -1,10, -1,13, -1,9, -1,12, -1,8,
-        -1,15, -1,11, -1,14, -1,10, -1,13, -1,9, -1,12, -1,8));
-      sss1 = _mm256_add_epi32(sss1, _mm256_madd_epi16(pix, mmk1));
+    unsigned int coefs_precision) {
+  // Interpolation horizontal pass processing together 4 vertical lines.
+  // - Input data format is RGBA with R,G,B,A being uint8, we can encode 4 values as a single uint32 value.
+  // - We split the size of weight vector for a given output index as a sum: K = n * 4 + m * 2 + k.
+  // - We load and process 4 weights values in a loop ("block 4") then we process 2 weights values
+  // in another loop ("block 2") and finally we process 1 weights value in the final loop ("block 1").
+
+  // Define shuffling masks (low/high) for num_channels 4
+  // Mask low casts lower half of each lane to epi16 and reorder RGBARGBA -> RRGGBBAA:
+  //   [r1 g1 b1 a1  r2 g2 b2 a2  ... | R1 G1 B1 A1  R2 G2 B2 A2 ... ] ->
+  //   [r1 0 r2 0  g1 0 g2 0  b1 0 b2 0  a1 0 a2 0 | R1 0 R2 0  G1 0 G2 0  B1 0 B2 0  A1 0 A2 0]
+  // Mask high casts upper half of each lane to epi16 and reorder RGBARGBA -> RRGGBBAA::
+  //   [ ... r3 g3 b3 a3  r4 g4 b4 a4 | ... R3 G3 B3 A3  R4 G4 B4 A4 ] ->
+  //   [r3 0 r4 0  g3 0 g4 0  b3 0 b4 0  a3 0 a4 0 | R3 0 R4 0  G3 0 G4 0  B3 0 B4 0  A3 0 A4 0]
+
+  const __m256i masks_low_high_c4[2] = {
+    _mm256_set_epi8(
+      -1, 7, -1, 3, -1, 6, -1, 2, -1, 5, -1, 1, -1, 4, -1, 0,
+      -1, 7, -1, 3, -1, 6, -1, 2, -1, 5, -1, 1, -1, 4, -1, 0
+    ),
+    _mm256_set_epi8(
+      -1, 15, -1, 11, -1, 14, -1, 10, -1, 13, -1, 9, -1, 12, -1, 8,
+      -1, 15, -1, 11, -1, 14, -1, 10, -1, 13, -1, 9, -1, 12, -1, 8
+    )
+  };
+
+  const auto mask_low = masks_low_high_c4[0];
+  const auto mask_high = masks_low_high_c4[1];
```

Comment:
Why have the array at all?

Reply:
In the next PR I add similar arrays for c3, but I agree there is no particular point of defining arrays vs 
```
masks_low_c4
masks_high_c4
```

Reply:
Please don't mark as resolved when you haven't fixed it.